spec/formatting.md

```diff
@@ -24,6 +34,14 @@ Formatting of a _message_ is defined by the following operations:
   The resolution of _text_ is rather straightforward,
   and is detailed under _literal resolution_.
 
+Formatter implementations are not required to expose
+the _expression resolution_ and _pattern selection_ operations to their users,
+or even use them in their internal processing,
+as long as the final _formatting_ result is made available to users
+and the observable behavior of the formatter matches that described here.
+
+_Attributes_ MUST NOT affect the processing or output of a _message_.
```

Comment:
This is too strong, since "processing" might reasonably include that used in translation, which is the whole point of attributes. Perhaps this is closer to what is intended:

```suggestion
_Attributes_ MUST NOT affect the _pattern selection_, _expression resolution_, or _formatting_ of a _message_.
```

Reply:
We discussed this previously in https://github.com/unicode-org/message-format-wg/pull/845#discussion_r1697081284, but that wasn't conclusive before that PR was merged.

I'd still prefer my suggestion from https://github.com/unicode-org/message-format-wg/pull/845#discussion_r1697159755:
```suggestion
_Attributes_ MUST NOT have any effect on the formatted output of a _message_,
or be made available to function implementations.
```

That ought to have the same effect, while not requiring us to enumerate all the parts of this document.

Would that work?

Reply:
I think so. Do we need to mention markup in addition to function implementation?

Reply:
No, because markup processing can't be customized within MF2, and therefore it's covered by "any effect on the formatted output".

Reply:
> +_Attributes_ MUST NOT have any effect on the formatted output of a _message_,
> +or be made available to function implementations.

This is slightly ambiguous. If you mean:

Attributes_ MUST NOT have any effect on the formatted output of a _message_,
AND
Attributes_ MUST NOT be made available to function implementations.

Then it would be better to say:
> +_Attributes_ MUST NOT have any effect on the formatted output of a _message_,
> +NOR be made available to function implementations.